File: 백준/Gold/2668. 숫자고르기/숫자고르기.py
n = int(input())
nums1= []
nums2 = []
visited =[False]*n
answer = []

for i in range(n):
  nums1.append(i+1)
  nums2.append(int(input()))

def dfs(curIndex, arr):
  global answer

  if visited[curIndex]:
    return
  
  arr.append(curIndex)
  visited[curIndex] = True

  val1 = list(map(lambda x:nums1[x], arr))
  val2 = list(map(lambda x:nums2[x], arr))
  val1.sort()
  val2.sort()
  
  isAllSame = True

  for i in range(len(val1)):
    if val1[i] != val2[i]:
      isAllSame = False
      break

  if isAllSame:
    for v in val1:
      if v not in answer:
        answer.append(v)

  dfs(nums2[curIndex]-1, arr)
  visited[curIndex] = False



for i in range(n):
  dfs(i, [])

answer.sort()
print(len(answer))
print("\n".join(list(map(str, answer))))

File: 백준/Gold/2668. 숫자고르기/숫자고르기.js
// const input = require("fs")
//   .readFileSync("example.txt", "utf8")
//   .trim()
//   .split("\n")
//   .map((line) => line.replace(/\r/, ""));

const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

// 첫번재줄에 숫자의 수
// 두번재줄부터 두번재줄 숫자들이 주어진다
// dfs로 1부터 방문지점을 다 돈다
// 그리고 up, down 집합이 완전히 일치하면 추가
const solution = (input) => {
  const n = Number(input[0]);
  const upArr = Array(n + 1)
    .fill(0)
    .map((v, index) => index);
  const downArr = [0];

  for (let i = 1; i <= n; i++) {
    downArr.push(Number(input[i]));
  }

  let answer = [];
  let upSample = [];
  let downSample = [];
  let visited = Array(n + 1).fill(false);

  const dfs = (value) => {
    if (visited[value]) return;
    visited[value] = true;
    upSample.push(value);
    downSample.push(downArr[value]);

    dfs(downArr[value]);
  };

  const set = new Set();

  for (let i = 1; i <= n; i++) {
    upSample = [];
    downSample = [];
    visited = Array(n + 1).fill(false);
    dfs(i);
    upSample.sort((a, b) => a - b);
    downSample.sort((a, b) => a - b);
    let isValid = true;

    for (let i = 0; i < upSample.length; i++) {
      if (upSample[i] !== downSample[i]) isValid = false;
    }

    if (isValid && !set.has(upSample.join(","))) {
      answer.push(...upSample);
      set.add(upSample.join(","));
    }
  }
  answer.sort((a, b) => a - b);
  console.log(answer.length);
  console.log(answer.join("\n"));
};

solution(input);
